src/black/lines.py

```diff
@@ -62,11 +62,21 @@ def append(self, leaf: Leaf, preformatted: bool = False) -> None:
         if token.COLON == leaf.type and self.is_class_paren_empty:
             del self.leaves[-2:]
         if self.leaves and not preformatted:
-            # Note: at this point leaf.prefix should be empty except for
-            # imports, for which we only preserve newlines.
-            leaf.prefix += whitespace(
-                leaf, complex_subscript=self.is_complex_subscript(leaf)
-            )
+            grandparent_leaf = leaf.parent.parent if leaf.parent else None
+
+            # It shouldn't add whitespaces after dot operators in expressions
```

Comment:
Looking at this again, shouldn't the logic you're adding be handled in the whitespace() function?

What if you remove the NUMBER check from nodes.py line 311, `if not prevp or prevp.type != token.NUMBER:`?